crmsh/ui_sbd.py

```diff
@@ -0,0 +1,551 @@
+                parameter_dict[key] = int(value)
+            # watchdog device parameter
+            elif key == "watchdog" and suffix == "device":
+                parameter_dict["watchdog-device"] = value
+            else:
+                raise self.SyntaxError(f"Unknown argument: {arg}")
+
+        watchdog_device = parameter_dict.get("watchdog-device")
+        parameter_dict["watchdog-device"] = watchdog.Watchdog.get_watchdog_device(watchdog_device)
+
+        logger.debug("Parsed arguments: %s", parameter_dict)
+        return parameter_dict
+
+    @staticmethod
+    def _adjust_timeout_dict(timeout_dict: dict) -> dict:
+        watchdog_timeout = timeout_dict.get("watchdog")
+        msgwait_timeout = timeout_dict.get("msgwait")
+        if watchdog_timeout and msgwait_timeout and msgwait_timeout < 2*watchdog_timeout:
+            logger.warning("It's recommended to set msgwait timeout >= 2*watchdog timeout")
+            return timeout_dict
+        if watchdog_timeout and not msgwait_timeout:
+            timeout_dict["msgwait"] = 2*watchdog_timeout
+            logger.info("No msgwait timeout specified, use 2*watchdog timeout: %s", 2*watchdog_timeout)
+            return timeout_dict
+        if msgwait_timeout and not watchdog_timeout:
+            watchdog_timeout = msgwait_timeout//2
+            timeout_dict["watchdog"] = watchdog_timeout
+            logger.info("No watchdog timeout specified, use msgwait timeout/2: %s", watchdog_timeout)
+            return timeout_dict
+
+    def _configure_diskbase(self, parameter_dict: dict):
+        '''
+        Configure disk-based SBD based on input parameters and runtime config
+        '''
+        update_dict = {}
+        watchdog_device = parameter_dict.get("watchdog-device")
+        if watchdog_device != self.watchdog_device_from_config:
+            update_dict["SBD_WATCHDOG_DEV"] = watchdog_device
+        timeout_dict = {k: v for k, v in parameter_dict.items() if k in self.TIMEOUT_TYPES}
+        is_subdict_timeout = utils.is_subdict(timeout_dict, self.device_meta_dict_runtime)
+
+        if is_subdict_timeout and not update_dict:
+            logger.info("No change in SBD configuration")
+            return
+
+        if not is_subdict_timeout:
+            timeout_dict = self._adjust_timeout_dict(timeout_dict)
+            # merge runtime timeout dict into parameter timeout dict without overwriting
+            timeout_dict = {**self.device_meta_dict_runtime, **timeout_dict}
+
+        watchdog_timeout = timeout_dict.get("watchdog")
+        if watchdog_timeout != self.watchdog_timeout_from_config:
+            update_dict["SBD_WATCHDOG_TIMEOUT"] = str(watchdog_timeout)
+
+        sbd_manager = sbd.SBDManager(
+            device_list_to_init=self.device_list_from_config,
+            timeout_dict=timeout_dict,
+            update_dict=update_dict
+        )
+        sbd_manager.init_and_deploy_sbd()
+        
+    def _configure_diskless(self, parameter_dict: dict):
```

Comment:
Same as the disk-based sbd, I expect to see TimeoutStartSec get updated along with `crm sbd configure watchdog-timeout=50`, for example.

Sounds like, the diskless bootstrap code doesn't do this too.

Reply:
Already changed, please have a try
```
# crm sbd  configure watchdog-timeout=100
INFO: Configuring diskless SBD
WARNING: Diskless SBD requires cluster with three or more nodes. If you want to use diskless SBD for 2-node cluster, should be combined with QDevice.
INFO: Update SBD_WATCHDOG_TIMEOUT in /etc/sysconfig/sbd: 100
INFO: Already synced /etc/sysconfig/sbd to all nodes
INFO: Restarting cluster service
INFO: BEGIN Waiting for cluster
...........                                                                                      INFO: END Waiting for cluster
INFO: Update SBD_DELAY_START in /etc/sysconfig/sbd: 211
INFO: Already synced /etc/sysconfig/sbd to all nodes
WARNING: "stonith-timeout" in crm_config is set to 251, it was 71

# crm sbd configure show
INFO: crm sbd configure show sysconfig
SBD_PACEMAKER=yes
SBD_STARTMODE=always
SBD_DELAY_START=211
SBD_WATCHDOG_DEV=/dev/watchdog0
SBD_WATCHDOG_TIMEOUT=100
SBD_TIMEOUT_ACTION=flush,reboot
SBD_MOVE_TO_ROOT_CGROUP=auto
SBD_SYNC_RESOURCE_STARTUP=yes
SBD_OPTS=

INFO: crm sbd configure show property
have-watchdog=true
stonith-enabled=true
stonith-watchdog-timeout=-1
stonith-timeout=251

INFO: systemctl show -p TimeoutStartUSec sbd.service --value
TimeoutStartUSec=253
```

Reply:
Still need the similar handling when the value is decreased. For example, 

```
# crm sbd  configure watchdog-timeout=100
# crm sbd  configure watchdog-timeout=10 # still, SBD_DELAY_START=211, stonith-timeout=251, TimeoutStartUSec=253
# crm sbd  configure watchdog-timeout=50 # still, SBD_DELAY_START=211, stonith-timeout=251, TimeoutStartUSec=253
```

Even, I expect to delete "/etc/systemd/system/sbd.service.d/sbd_delay_start.conf" when TimeoutStartUSec <= 90

Reply:
Take `SBD_DELAY_START` for an example, current our algo is, if expected_value > configured _value, update, see https://github.com/ClusterLabs/crmsh/blob/master/crmsh/sbd.py#L232-L235
Do you mean we changed as `if expected_value != configuted_value`, update?

Reply:
Good argument! You help me to recall my vague memory, in the past we let the algorithm adjust the value only if it can go wrong badly, and leave the value by assuming they might be manually configured for reasons . 

I wonder if makes more sense to refactor and add a more complex flag to control the levels to adjust those related values. Em, yes, we need think twice. Keep debating ... no rush.

Apparently, some related improvements as below is reasonable to fix now. `crm sbd purge` should cleanup systemd drop-in files regardless if sbd is the last stonith device. Same for stonith-watchdog-timeout.
 
  /etc/systemd/system/sbd.service.d/sbd_delay_start.conf
  /run/systemd/system/sbd.service.d/sbd_delay_start_disabled.conf

Reply:
> Good argument! You help me to recall my vague memory, in the past we let the algorithm adjust the value only if it can go wrong badly, and leave the value by assuming they might be manually configured for reasons .
> 
> I wonder if makes more sense to refactor and add a more complex flag to control the levels to adjust those related values. Em, yes, we need think twice. Keep debating ... no rush.

Yeah, maybe raise an upstream issue or create a confluence/jira page?

Reply:
> Apparently, some related improvements as below is reasonable to fix now. `crm sbd purge` should cleanup systemd drop-in files regardless if sbd is the last stonith device. Same for stonith-watchdog-timeout.
> 
> /etc/systemd/system/sbd.service.d/sbd_delay_start.conf /run/systemd/system/sbd.service.d/sbd_delay_start_disabled.conf

Add commit: Dev: sbd: Remove sbd delay start related diretories when running sbd purge